frontend/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx

```diff
@@ -43,6 +45,87 @@ export interface Props {
   allowHTML: boolean
 }
 
+function createAnchorFromText(text: string | null): string {
+  const newAnchor = text
+    ?.toLowerCase()
+    .split(/[^A-Za-z0-9]/)
+    .filter(Boolean)
+    .join("-")
+  return newAnchor || ""
+}
+
+// wrapping in `once` ensures we only scroll once
+const scrollNodeIntoView = once((node: HTMLElement): void => {
+  node.scrollIntoView(true)
+})
+
+function HeadingWithAnchor({
+  tag,
+  anchor: propsAnchor,
+  children,
+}: any): ReactElement {
+  const [elementId, setElementId] = React.useState(propsAnchor)
+  const [target, setTarget] = React.useState<HTMLElement | null>(null)
+
+  const {
+    addReportFinshedHandler,
+    removeReportFinshedHandler,
+  } = React.useContext(PageLayoutContext)
+
+  const onReportFinished = React.useCallback(() => {
+    if (target !== null) {
+      // wait a bit for everything on page to finish loading
+      window.setTimeout(() => {
+        scrollNodeIntoView(target)
+      }, 300)
+    }
+  }, [target])
+
+  React.useEffect(() => {
+    addReportFinshedHandler(onReportFinished)
+    return () => {
+      removeReportFinshedHandler(onReportFinished)
+    }
+  }, [addReportFinshedHandler, removeReportFinshedHandler, onReportFinished])
+
+  const ref = React.useCallback(
+    node => {
+      if (node === null) return
+
+      const anchor = propsAnchor || createAnchorFromText(node.textContent)
+      setElementId(anchor)
+      if (window.location.hash.slice(1) === anchor) {
+        setTarget(node)
+      }
+    },
+    [propsAnchor]
+  )
+
+  return React.createElement(tag, { ref, id: elementId }, children)
+}
+
+function CustomHeading({ level, children }: any): ReactElement {
+  return <HeadingWithAnchor tag={`h${level}`}>{children}</HeadingWithAnchor>
+}
+
+function CustomParsedHtml(props: any): ReactElement {
+  const {
+    element: { type, props: elementProps },
+  } = props
+
+  const headingElements = ["h1", "h2", "h3", "h4", "h5", "h6"]
+  if (!headingElements.includes(type)) {
+    return (ReactMarkdown.renderers.parsedHtml as any)(props)
```

Comment:
and here

Reply:
I casted to `any` here because all we're doing here is "call ReactMarkdown's original renderer," and the ReactMarkdown's types are being weird. I think it should be fine, this isn't our own code. I added a comment.

Reply:
I casted to `any` here because all we're doing here is "call ReactMarkdown's original renderer," and ReactMarkdown's types are being weird. I think it should be fine, this isn't our own code. I added a comment.

Reply:
Cool 👍